source/postprocess/viscous_dissipation_per_composition.cc

```diff
@@ -0,0 +1,177 @@
+
+      const unsigned int n_q_points = quadrature_formula.size();
+
+      FEValues<dim> fe_values (this->get_mapping(),
+                               this->get_fe(),
+                               quadrature_formula,
+                               update_values   |
+                               update_quadrature_points |
+                               update_gradients |
+                               update_JxW_values);
+
+      std::vector<double> local_dissipation_integral (this->n_compositional_fields());
+
+      std::vector<double> compositional_values(n_q_points);
+
+      // the values of the compositional fields are stored as blockvectors for each field
+      // we have to extract them in this structure
+      std::vector<std::vector<double> > prelim_composition_values (this->n_compositional_fields(),
+                                                                   std::vector<double> (n_q_points));
+
+      typename MaterialModel::Interface<dim>::MaterialModelInputs in(n_q_points,
+                                                                     this->n_compositional_fields());
+      typename MaterialModel::Interface<dim>::MaterialModelOutputs out(n_q_points,
+                                                                       this->n_compositional_fields());
+
+      for (const auto &cell : this->get_dof_handler().active_cell_iterators())
+        if (cell->is_locally_owned())
+          {
+            fe_values.reinit (cell);
+            in.reinit(fe_values, cell, this->introspection(), this->get_solution());
+
+            this->get_material_model().fill_additional_material_model_inputs(in, this->get_solution(), fe_values, this->introspection());
+
+            this->get_material_model().evaluate(in, out);
+
+            for (unsigned int c = 0; c<this->n_compositional_fields(); ++c)
+              {
+                for (unsigned int q = 0; q < n_q_points; ++q)
+                  {
+                    if (in.composition[q][c] >= 0.5)
+                      {
+                        // Viscous dissipation D:
+                        // D = 1/2 * volume_integral(deviatoric_stress*deviatoric_strain_rate)
+                        //   = 1/2 * volume_integral(2*eta*deviatoric_strain_rate*deviatoric_strain_rate)
+                        //   = volume_integral(eta*deviatoric_strain_rate*deviatoric_strain_rate)
+                        const SymmetricTensor<2, dim> deviatoric_strain_rate =
+                          (this->get_material_model().is_compressible()
+                           ? in.strain_rate[q] - 1. / 3. * trace(in.strain_rate[q]) * unit_symmetric_tensor<dim>()
+                           : in.strain_rate[q]);
+                        local_dissipation_integral[c] += (out.viscosities[q] * deviatoric_strain_rate) *
+                                                         deviatoric_strain_rate * fe_values.JxW(q);
+                      }
+                  }
+              }
+          }
```

Comment:
If you write it like this you save a lot of time, and can combine both postprocessors into this one:

```suggestion
                     const unsigned int n_compositional_fields = this->n_compositional_fields();     
                for (unsigned int q = 0; q < n_q_points; ++q)
                  {
                        // Viscous dissipation D:
                        // D = 1/2 * volume_integral(deviatoric_stress*deviatoric_strain_rate)
                        //   = 1/2 * volume_integral(2*eta*deviatoric_strain_rate*deviatoric_strain_rate)
                        //   = volume_integral(eta*deviatoric_strain_rate*deviatoric_strain_rate)
                        const SymmetricTensor<2, dim> deviatoric_strain_rate =
                          (this->get_material_model().is_compressible()
                           ? in.strain_rate[q] - 1. / 3. * trace(in.strain_rate[q]) * unit_symmetric_tensor<dim>()
                           : in.strain_rate[q]);
                        const double local_dissipation = (out.viscosities[q] * deviatoric_strain_rate) *
                                                         deviatoric_strain_rate * fe_values.JxW(q);
                        

                         // for this to work make the vector one element longer, and store the overall integral in
                         // the last element.                                                       
                        local_dissipation_integral[n_compositional_fields] += local_dissipation;
                      
                                  for (unsigned int c = 0; c<n_compositional_fields; ++c)
                                                      if (in.composition[q][c] >= 0.5)
                                                        local_dissipation_integral[c] += local_dissipation;
              }
          }
```